pkg/icingadb/history/sync.go

```diff
@@ -32,274 +35,366 @@ func NewSync(db *icingadb.DB, redis *icingaredis.Client, logger *zap.SugaredLogg
+			i, stage := i, stage
 			g.Go(func() error {
-				defer close(succeeded)
-				return s.db.UpsertStreamed(ctx, struct2db, succeeded)
+				return stage(ctx, s, key, ch[i], ch[i+1])
 			})
 		}
 
-		g.Go(s.xRead(ctx, redis2structs, stream))
-		g.Go(s.cleanup(ctx, hs, insertedMessages, messageProgress, messageProgressMtx, stream))
+		g.Go(func() error {
+			return s.deleteFromRedis(ctx, key, ch[len(pipeline)])
+		})
 	}
 
 	return g.Wait()
 }
 
-// xRead reads from the Redis stream and broadcasts the data to redis2structs.
-func (s Sync) xRead(ctx context.Context, redis2structs []chan<- redis.XMessage, stream string) func() error {
-	return func() error {
-		defer func() {
-			for _, r2s := range redis2structs {
-				close(r2s)
-			}
-		}()
+// readFromRedis is the first stage of the history sync pipeline. It reads the history stream from Redis
+// and feeds the history entries into the next stage.
+func (s Sync) readFromRedis(ctx context.Context, key string, output chan<- redis.XMessage) error {
+	defer close(output)
 
-		xra := &redis.XReadArgs{
-			Streams: []string{stream, "0-0"},
-			Count:   bulkSize,
-			Block:   10 * time.Second,
-		}
+	xra := &redis.XReadArgs{
+		Streams: []string{"icinga:history:stream:" + key, "0-0"},
+		Count:   bulkSize,
+		Block:   10 * time.Second,
+	}
 
-		for {
-			cmd := s.redis.XRead(ctx, xra)
-			streams, err := cmd.Result()
+	for {
+		cmd := s.redis.XRead(ctx, xra)
+		streams, err := cmd.Result()
 
-			if err != nil && err != redis.Nil {
-				return icingaredis.WrapCmdErr(cmd)
-			}
+		if err != nil && err != redis.Nil {
+			return icingaredis.WrapCmdErr(cmd)
+		}
 
-			for _, stream := range streams {
-				for _, message := range stream.Messages {
-					xra.Streams[1] = message.ID
+		for _, stream := range streams {
+			for _, message := range stream.Messages {
+				xra.Streams[1] = message.ID
 
-					for _, r2s := range redis2structs {
-						select {
-						case <-ctx.Done():
-							return ctx.Err()
-						case r2s <- message:
-						}
-					}
+				select {
+				case output <- message:
+				case <-ctx.Done():
+					return ctx.Err()
 				}
 			}
 		}
 	}
 }
 
-// structifyStream structifies from redis2struct to struct2db.
-func structifyStream(
-	ctx context.Context, structifier structify.MapStructifier, redis2struct <-chan redis.XMessage,
-	struct2db chan<- contracts.Entity, rowIds map[contracts.Entity]string, rowIdsMtx *sync.Mutex,
-) func() error {
-	return func() error {
-		defer close(struct2db)
-
-		for {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case message, ok := <-redis2struct:
-				if !ok {
-					return nil
+// deleteFromRedis is the last stage of the history sync pipeline. It receives history entries from the second to last
+// pipeline stage and then deletes the stream entry from Redis as all pipeline stages successfully processed the entry.
+func (s Sync) deleteFromRedis(ctx context.Context, key string, input <-chan redis.XMessage) error {
```

Comment:
Hmm, `Bulker` seems to implement mostly the same logic, but operates on `chan interface{}` so I'd either have to use `interface{}` in most places that use `redis.XMessage` right now or have another goroutine that just converts from `chan redis.XMessage` to `chan interface{}`.

Reply:
Yeah, that's true. We already introduced `EntityBulker` because of that. Is something like this worth it?

- Introduce `type Bulkable interface {}`
- s/interface{}/Bulkable
- Embed interfaces
- Remove `EntityBulker`
- Use `Bulker` here

But then we would have to use type assertion everywhere which is not so nice.

Reply:
Doesn't change anything, still would need type assertions everywhere. Generics would help.